codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,319 @@
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler.UpgradeEvent;
+import io.netty.util.AsciiString;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class Http2MultiplexCodec extends ChannelHandlerAppender {
+    private static final Http2FrameLogger logger
+        = new Http2FrameLogger(INFO, Http2MultiplexCodec.class);
+
+    private final boolean server;
+    private final ChannelHandler streamHandler;
+    private final EventLoopGroup streamGroup;
+    private final Http2ConnectionHandler http2Handler;
+    private ChannelHandlerContext ctx;
+    private final List<StreamInfo> streamsToFireChildReadComplete = new ArrayList<StreamInfo>();
+
+    /** {@code streamHandler} should be {@link ChannelHandler.Sharable}. */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+        this(server, streamHandler, null);
+    }
+
+    /** {@code streamHandler} should be {@link ChannelHandler.Sharable}. */
+    public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler,
+            EventLoopGroup streamGroup) {
+        super(false /* selfRemoval */);
+        this.server = server;
+        this.streamHandler = streamHandler;
+        this.streamGroup = new io.netty.channel.DefaultEventLoopGroup(); //streamGroup;
+        http2Handler = new Http2ConnectionHandler(new DefaultHttp2Connection(server),
+                new Http2InboundFrameLogger(new DefaultHttp2FrameReader(), logger),
+                new Http2OutboundFrameLogger(new DefaultHttp2FrameWriter(), logger),
+                new FrameListener());
+        http2Handler.connection().addListener(new ConnectionListener());
+        add(http2Handler);
+    }
+
+    Http2ConnectionHandler connectionHandler() {
+        return http2Handler;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = ctx;
+        super.handlerAdded(ctx);
+    }
+
+    /**
+     * Handles the cleartext HTTP upgrade event. If an upgrade occurred, sends a simple response via
+     * HTTP/2 on stream 1 (the stream specifically reserved for cleartext HTTP upgrade).
+     */
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (!(evt instanceof UpgradeEvent)) {
+            super.userEventTriggered(ctx, evt);
+        }
+
+        UpgradeEvent upgrade = (UpgradeEvent) evt;
+        FullHttpRequest request = upgrade.upgradeRequest();
+        Http2Stream stream = http2Handler.connection().requireStream(1);
+        // The stream was already made active, but ctx may have been null so it wasn't initialized.
+        new ConnectionListener().onStreamActive(stream);
```

Comment:
Well, we are actually guaranteed to have an instance of ConnectionListener already created, so we could just keep a reference to it. However, honestly, this code is called only once per connection, and then only for plain-text. I have trouble caring.